src/group_theory/complement.lean

```diff
@@ -0,0 +1,303 @@
+function.surjective.has_right_inverse quotient.surjective_quotient_mk', f := classical.some hf in
+⟨⟨set.range f, mem_right_transverals_iff_bijective.mpr ⟨by
+{ rintros ⟨_, q₁, rfl⟩ ⟨_, q₂, rfl⟩ hg,
+  rw [subtype.ext_iff, subtype.coe_mk, subtype.coe_mk,
+      ←classical.some_spec hf q₁, ←classical.some_spec hf q₂],
+  exact congr_arg f hg }, λ q, ⟨⟨f q, q, rfl⟩, classical.some_spec hf q⟩⟩⟩⟩
+
+lemma is_complement_of_disjoint [fintype G] [fintype H] [fintype K]
+  (h1 : fintype.card H * fintype.card K = fintype.card G)
+  (h2 : disjoint H K) :
+  is_complement (H : set G) (K : set G) :=
+begin
+  refine function.bijective.exists_unique ((fintype.bijective_iff_injective_and_card _).mpr
+    ⟨λ x y h, _, (fintype.card_prod H K).trans h1⟩),
+  rw [←eq_inv_mul_iff_mul_eq, ←mul_assoc, ←mul_inv_eq_iff_eq_mul] at h,
+  change ↑(x.2 * y.2⁻¹) = ↑(x.1⁻¹ * y.1) at h,
+  rw [prod.ext_iff, ←@inv_mul_eq_one H _ x.1 y.1, ←@mul_inv_eq_one K _ x.2 y.2, subtype.ext_iff,
+      subtype.ext_iff, coe_one, coe_one, h, and_self, ←mem_bot, ←h2.eq_bot, mem_inf],
+  exact ⟨subtype.mem ((x.1)⁻¹ * (y.1)), (congr_arg (∈ K) h).mp (subtype.mem (x.2 * (y.2)⁻¹))⟩,
+end
+
+lemma is_complement_of_coprime [fintype G] [fintype H] [fintype K]
+  (h1 : fintype.card H * fintype.card K = fintype.card G)
+  (h2 : nat.coprime (fintype.card H) (fintype.card K)) :
+  is_complement (H : set G) (K : set G) :=
+is_complement_of_disjoint h1 (disjoint_iff.mpr (inf_eq_bot_of_coprime h2))
+
+section schur_zassenhaus
+
+@[to_additive] instance : mul_action G (left_transversals (H : set G)) :=
+{ smul := λ g T, ⟨left_coset g T, mem_left_transversals_iff_exists_unique_inv_mul_mem.mpr (λ g', by
+  { obtain ⟨t, ht1, ht2⟩ := mem_left_transversals_iff_exists_unique_inv_mul_mem.mp T.2 (g⁻¹ * g'),
+    simp_rw [←mul_assoc, ←mul_inv_rev] at ht1 ht2,
+    refine ⟨⟨g * t, mem_left_coset g t.2⟩, ht1, _⟩,
+    rintros ⟨_, t', ht', rfl⟩ h,
+    exact subtype.ext ((mul_right_inj g).mpr (subtype.ext_iff.mp (ht2 ⟨t', ht'⟩ h))) })⟩,
+  one_smul := λ T, subtype.ext (one_left_coset T),
+  mul_smul := λ g g' T, subtype.ext (left_coset_assoc ↑T g g').symm }
+
+lemma smul_symm_apply_eq_mul_symm_apply_inv_smul
+  (g : G) (α : left_transversals (H : set G)) (q : quotient_group.quotient H) :
+  ↑((equiv.of_bijective _ (mem_left_transverals_iff_bijective.mp (g • α).2)).symm q) =
+    g * ((equiv.of_bijective _ (mem_left_transverals_iff_bijective.mp α.2)).symm
+      (g⁻¹ • q : quotient_group.quotient H)) :=
+begin
+  let w := (equiv.of_bijective _ (mem_left_transverals_iff_bijective.mp α.2)),
+  let y := (equiv.of_bijective _ (mem_left_transverals_iff_bijective.mp (g • α).2)),
+  change ↑(y.symm q) = ↑(⟨_, mem_left_coset g (subtype.mem _)⟩ : (g • α).1),
+  refine subtype.ext_iff.mp (y.symm_apply_eq.mpr _),
+  change q = g • (w (w.symm (g⁻¹ • q : quotient_group.quotient H))),
+  rw [equiv.apply_symm_apply, ←mul_smul, mul_inv_self, one_smul],
+end
+
+@[to_additive] instance [h : is_commutative H (*)] : comm_group H :=
```

Comment:
Perhaps we already resolved this in a zulip thread, but this instance looks dangerous to me, as it forms a loop.

Reply:
My impression from the zulip thread (https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Abelian.20Subgroup) was that having this instance for `H : subgroup G` wasn't a problem, but I could be wrong

Reply:
I can't be sure whether it's a problem, but it certainly forms a loop. In principle, the following could happen:

* Lean wants `comm_group H` and finds this instance. It now needs to find `is_commutative H (*)`
* Lean wants `is_commutative H (*)` and finds `comm_semigroup.to_is_commutative`. It now needs to find `comm_semigroup H`.
* Lean wants `comm_semigroup H`, and finds `comm_monoid.to_comm_semigroup`. It now needs to find `comm_monoid H`.
* Lean wants `comm_monoid H`, and finds `comm_group.to_comm_monoid`. It now needs to find `comm_group H`.
* Oops, we have a loop.

Perhaps @digama0 can confirm whether this actually happens

Reply:
Ah, I see. So what's the fix?

Reply:
Making a new `subgroup.is_commutative` typeclass is all I can think of